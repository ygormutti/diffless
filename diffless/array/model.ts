import { JSIN } from '../jsin';
import { DiffLevel, Document, Equals, Excerpt, Location, Weigh } from '../model';

export type Tokenizer = (document: Document) => Token[];

export interface ArrayDiffOptions<TExcerpt extends Excerpt> {
    level: DiffLevel;
    excerptMapper: ExcerptMapper<TExcerpt>;
    similarityThreshold: number;
    equals?: Equals<TExcerpt>;
    weigh?: Weigh<TExcerpt>;
}

export type ExcerptMapper<TExcerpt extends Excerpt> = (document: Document) => TExcerpt[];

@JSIN.enabled
export class Token extends Excerpt {
    constructor(
        location: Location,
        content: string,
        readonly type: number,
    ) {
        super(content, location);
    }

    equals(other: Token): boolean {
        return !(other instanceof ValuedToken) && this.type === other.type;
    }

    static equals(a: Token, b: Token) {
        return a.equals(b);
    }
}

@JSIN.enabled
export class ValuedToken<TTokenValue> extends Token {
    constructor(
        location: Location,
        content: string,
        type: number,
        readonly value: TTokenValue,
        readonly valueEquals: Equals<TTokenValue>,
    ) {
        super(location, content, type);
    }

    equals(other: Token): boolean {
        return other instanceof ValuedToken
            && this.type === other.type
            && this.valueEquals(this.value, other.value);
    }
}
