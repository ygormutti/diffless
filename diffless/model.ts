import { JSIN } from './jsin';

/**
 * Model classes and interfaces.
 *
 * Design and docs inspired by: https://microsoft.github.io/language-server-protocol/specification
 */

/**
 * The URI of a document
 */
export type DocumentURI = string;

/**
 * Function type that checks if two objects are equal according to some criteria
 */
export type Equals<T> = (a: T, b: T) => boolean;

/**
 * Function type that returns the object weigh
 */
export type Weigh<T> = (obj: T) => number;

/**
 * The end of line sequence adopted for internal use
 */
export const EOL = '\n';

/**
 * A type which instances have a natural total order
 */
export interface Comparable<T> {
    compareTo(other: T): number;
}

/**
 * Position in a text document expressed as one-based line and one-based character offset.
 *
 * A position is between two characters like an ‘insert’ cursor in a editor.
 */
@JSIN.enabled
export class Position implements Comparable<Position> {
    constructor(
        /**
         * Line position in a document (one-based).
         */
        readonly line: number,

        /**
         * Character offset on a line in a document (one-based). Assuming that the line is
         * represented as a string, the `character` value represents the gap between the
         * `character` and `character + 1`.
         *
         * If the character value is greater than the line length it defaults back to the
         * line length.
         */
        readonly character: number,
    ) { }

    get lineOffset() {
        return this.line - 1;
    }

    get characterOffset() {
        return this.character - 1;
    }

    get code() {
        return `${this.line}:${this.character}`;
    }

    equals(other: Position) {
        return this.compareTo(other) === 0;
    }

    compareTo(other: Position): number {
        let value = this.line - other.line;
        if (!value) { value = this.character - other.character; }
        return value;
    }
}

/**
 * A range in a text document expressed as (one-based) start and end positions.
 *
 * A range is comparable to a selection in an editor. Therefore the end position is exclusive.
 * If you want to specify a range that contains a line including the line ending character(s)
 * then use an end position denoting the start of the next line.
 */
@JSIN.enabled
export class Range implements Comparable<Range> {
    constructor(
        /**
         * The range's start position.
         */
        readonly start: Position,

        /**
         * The range's end position.
         */
        readonly end: Position,
    ) { }

    get code() {
        return `[${(this.start.code)}; ${this.end.code})`;
    }

    compareTo(other: Range): number {
        let value = this.start.compareTo(other.start);
        if (!value) { value = this.end.compareTo(other.end); }
        return value;
    }

    static forLine(line: number) {
        return new Range(new Position(line, 1), new Position(line + 1, 1));
    }

    static forCharacter(position: Position, content: string) {
        const end = content === EOL ?
            new Position(position.line + 1, 1) :
            new Position(position.line, position.character + 1);
        return new Range(position, end);
    }
}

@JSIN.enabled
export class Location {
    constructor(
        readonly uri: DocumentURI,
        readonly range: Range,
    ) { }

    get code() {
        return `${this.range.code} @ ${this.uri}`;
    }

    static forLine(documentURI: DocumentURI, line: number) {
        return new Location(documentURI, Range.forLine(line));
    }

    static forCharacter(documentURI: DocumentURI, position: Position, content: string) {
        return new Location(documentURI, Range.forCharacter(position, content));
    }
}

/**
 * The unit of comparison of an atomic diff tool
 */
@JSIN.enabled
export class Atom {
    constructor(
        readonly content: string,
        readonly location: Location,
    ) { }

    get start() {
        return this.location.range.start;
    }

    get end() {
        return this.location.range.end;
    }

    static contentLength(atom: Atom) {
        return atom.content.length;
    }

    static sameContent(a: Atom, b: Atom) {
        return a.content === b.content;
    }
}

/**
 * A line from a text document
 */
@JSIN.enabled
export class Line extends Atom {
    constructor(
        content: string,
        documentURI: DocumentURI,
        readonly line: number,
    ) {
        super(content, Location.forLine(documentURI, line));
    }
}

/**
 * A single character from a text document
 */
@JSIN.enabled
export class Character extends Atom {
    constructor(
        content: string,
        documentURI: DocumentURI,
        readonly position: Position,
    ) {
        super(content, Location.forCharacter(documentURI, position, content));
    }
}

/**
 * The default content type for text documents
 */
export const DEFAULT_CONTENT_TYPE = 'text/plain';

/**
 * A text document
 */
@JSIN.enabled
export class Document {
    readonly lines: Line[];
    readonly characters: Character[];

    constructor(
        readonly uri: DocumentURI,
        readonly content: string,
        readonly contentType: string = DEFAULT_CONTENT_TYPE,
    ) {
        const linesContents = normalizeEOLs(content).split(EOL);
        this.lines = buildLines(linesContents, uri);
        this.characters = buildCharacters(this.lines, uri);
    }

    getRange(range: Range): string {
        const { start, end } = range;
        const lines = this.lines.filter((_, i) => start.lineOffset <= i && i <= end.lineOffset);

        if (lines.length === 1) {
            return lines[0].content.substring(start.characterOffset, end.characterOffset);
        }

        let rangeString = '';
        for (let lineOffset = 0; lineOffset < lines.length; lineOffset++) {
            const line = lines[lineOffset].content;
            let startOffset = 0;
            let endOffset = line.length;
            if (lineOffset === 0) {
                // first line
                startOffset = start.characterOffset;
            } else if (lineOffset === lines.length - 1) {
                // last line
                endOffset = end.characterOffset;
            }

            rangeString += line.substring(startOffset, Math.min(line.length, endOffset));
        }
        return rangeString;
    }
}

function buildLines(linesContents: string[], documentURI: DocumentURI): Line[] {
    return linesContents.map((v, i) => new Line(v + EOL, documentURI, i + 1));
}

function buildCharacters(lines: Line[], documentURI: DocumentURI): Character[] {
    const characters: Character[] = [];
    for (const [lineOffset, line] of lines.entries()) {
        const lineNumber = lineOffset + 1;
        for (const [characterOffset, character] of Array.from(line.content).entries()) {
            const position = new Position(lineNumber, characterOffset + 1);
            characters.push(new Character(character, documentURI, position));
        }
    }
    return characters;
}

function normalizeEOLs(content: string): string {
    return content.replace('\r\n', EOL).replace('\r', EOL);
}

export enum EditOperation {
    Add,
    Delete,
    Replace,
    Move,
    Copy,
    Rename,
}

export enum DiffLevel {
    Binary,
    Textual,
    Lexical,
    Syntactic,
    Semantic,
    DataFlow,
}

@JSIN.enabled
export class DiffItem {
    constructor(
        readonly level: DiffLevel,
        readonly left?: Location,
        readonly right?: Location,
    ) { }
}

@JSIN.enabled
export class Edit extends DiffItem implements Comparable<Edit> {
    constructor(
        level: DiffLevel,
        readonly operation: EditOperation,
        left?: Location,
        right?: Location,
    ) {
        super(level, left, right);
    }

    get code() {
        return `${DiffLevel[this.level]} ${EditOperation[this.operation].toLowerCase()}, ` +
            (this.left ? `L: ${this.left!.code}` : '') +
            (this.left && this.right ? ', ' : '') +
            (this.right ? `R: ${this.right!.code}` : '');
    }

    compareTo(other: Edit): number {
        let value = other.level - this.level;
        if (!value) { value = other.operation - this.operation; }
        if (!value && this.right && other.right) { value = this.right.range.compareTo(other.right.range); }
        if (!value && this.left && other.left) { value = this.left.range.compareTo(other.left.range); }
        return value;
    }
}

@JSIN.enabled
export class Similarity extends DiffItem { }

@JSIN.enabled
export class DocumentDiff {
    constructor(
        readonly left: Document,
        readonly right: Document,
        readonly edits: Edit[],
        readonly similarities: Similarity[],
    ) { }
}

export type DiffTool = (left: Document, right: Document) => DocumentDiff;

export type DiffToolFactory = (options: any) => DiffTool;
