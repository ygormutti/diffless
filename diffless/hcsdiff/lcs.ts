import { Equals } from '../model';

/**
 * Longest Common Substring model and implementation
 */

/**
 * Function type that finds the LCS between two arrays
 */
export type LCS = <TItem>(equal: Equals<TItem>, left: TItem[], right: TItem[]) => LCSResult<TItem>;

/**
 * The Longest Common Substring of two arrays an its offset in both arrays
 */
export class LCSResult<TItem> {
    constructor(
        readonly lcs: TItem[],
        readonly leftOffset: number,
        readonly rightOffset: number,
    ) { }
}

/**
 * Solves LCS by using dynamic programming in O(left.length * right.length) time and space
 *
 * Based on:
 * https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Longest_common_substring#JavaScript
 *
 * @param left left array
 * @param right right array
 */
export function dynamicProgrammingLCS<TItem>(
    equal: Equals<TItem>,
    left: TItem[],
    right: TItem[],
): LCSResult<TItem> {
    if (!left || !left.length || !right || !right.length) {
        return new LCSResult([], 0, 0);
    }

    const table = getInitialTable(left.length, right.length);

    let lcs: TItem[] = [];
    let lcsLength = 0;
    let lcsLeftOffset = 0;
    let csLeftOffset = 0;
    let csRightOffset = 0;

    for (let i = 0; i < left.length; i++) {
        for (let j = 0; j < right.length; j++) {
            if (!equal(left[i], right[j])) {
                table[i][j] = 0;
            } else {
                if ((i === 0) || (j === 0)) {
                    table[i][j] = 1;
                } else {
                    table[i][j] = 1 + table[i - 1][j - 1];
                }

                const csLength = table[i][j];
                if (csLength > lcsLength) {
                    lcsLength = csLength;
                    csLeftOffset = i - csLength + 1;
                    csRightOffset = j - csLength + 1;

                    if (lcsLeftOffset === csLeftOffset) {
                        // same LCS
                        lcs.push(left[i]);
                    } else {
                        // new LCS
                        lcsLeftOffset = csLeftOffset;
                        lcs = left.slice(lcsLeftOffset, lcsLeftOffset + lcsLength);
                    }
                }
            }
        }
    }

    return new LCSResult(lcs, csLeftOffset, csRightOffset);
}

function getInitialTable(rows: number, columns: number): number[][] {
    const table = new Array(rows);
    for (let i = 0; i < rows; i++) {
        const row = new Array(columns);
        row.fill(0);
        table[i] = row;
    }
    return table;
}
