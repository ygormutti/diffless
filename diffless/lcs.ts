/**
 * Function that finds the LCS between two arrays
 */
export interface LCS {
    <T>(equal: Equal<T>, left: T[], right: T[]): LCSResult<T>;
}

/**
 * Function that checks if two objects are equal according to some criteria
 */
export interface Equal<T> {
    (left: T, right: T): boolean;
}

/**
 * The Longest Common Subsequence of two arrays an its offset in both arrays
 */
export class LCSResult<T> {
    constructor(
        readonly lcs: T[],
        readonly leftOffset: number,
        readonly rightOffset: number,
    ) { }
}

function getInitialTable(rows: number, columns: number): number[][] {
    const table = new Array(rows);
    for (let i = 0; i < rows; i++) {
        const row = new Array(columns);
        row.fill(0);
        table[i] = row;
    }
    return table;
}

/**
 * Solves LCS by using dynamic programming in O(left.length * right.length) time and space
 * Based on: https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Longest_common_substring#JavaScript
 * @param left left sequence
 * @param right right sequence
 */
export function dynamicProgrammingLCS<T>(equal: Equal<T>, left: T[], right: T[]): LCSResult<T> {
    if (!left || !left.length || !right || !right.length) {
        return new LCSResult([], 0, 0);
    }

    const table = getInitialTable(left.length, right.length);

    let lcs: T[] = [];
    let lcsLength = 0;
    let lcsLeftOffset = 0;
    let lcsRightOffset = 0;
    let csLeftOffset = 0;
    let csRightOffset = 0;

    for (let i = 0; i < left.length; i++) {
        for (let j = 0; j < right.length; j++) {
            if (!equal(left[i], right[j])) {
                table[i][j] = 0;
            } else {
                if ((i === 0) || (j === 0)) {
                    table[i][j] = 1;
                } else {
                    table[i][j] = 1 + table[i - 1][j - 1];
                }

                const csLength = table[i][j];
                if (csLength > lcsLength) {
                    lcsLength = csLength;
                    csLeftOffset = i - csLength + 1;
                    csRightOffset = j - csLength + 1;

                    if (lcsLeftOffset === csLeftOffset) {
                        // same LCS
                        lcs.push(left[i]);
                    } else {
                        // new LCS
                        lcsLeftOffset = csLeftOffset;
                        lcsRightOffset = csRightOffset;
                        lcs = left.slice(lcsLeftOffset, lcsLeftOffset + lcsLength);
                    }
                }
            }
        }
    }

    return new LCSResult(lcs, csLeftOffset, csRightOffset);
}
