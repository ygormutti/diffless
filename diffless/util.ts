import { Edit, Location, Position } from './model';

export class EditIndex {
    private index: Edit[][][];

    constructor() {
        this.index = [];
    }

    add(location: Location, change: Edit) {
        const { start, end } = location.range;
        this.addToPosition(start, change);
        this.addToPosition(end, change);
    }

    private addToPosition(position: Position, change: Edit) {
        const { line, character } = position;
        if (!this.index[line]) { this.index[line] = []; }

        let changes = this.index[line][character];
        if (!changes) {
            changes = [];
            this.index[line][character] = changes;
        }

        changes.push(change);

        // sort by level
        changes.sort((a, b) => {
            return ((a.level * 100) - (b.level * 100)) + (a.type - b.type);
        });
    }

    get(position: Position): Edit[] {
        const { line, character } = position;
        if (!this.index[line] || !this.index[line][character]) { return []; }
        return this.index[line][character];
    }
}

export function buildEditIndexes(changes: Edit[]): [EditIndex, EditIndex] {
    const leftChangeIndex = new EditIndex();
    const rightChangeIndex = new EditIndex();
    changes.forEach(change => {
        const { left, right } = change;
        if (left) { leftChangeIndex.add(left, change); }
        if (right) { rightChangeIndex.add(right, change); }
    });
    return [leftChangeIndex, rightChangeIndex];
}

export function sortEdits(changes: Edit[]): void {
    changes.sort((a: Edit, b: Edit) => {
        let value = a.level - b.level;
        if (!value) {
            value = a.type - b.type;
        }
        return value;
    });
}

export function intEnumKeys(E: any): string[] {
    return Object.keys(E).filter(k => typeof E[k as any] === 'number');
}

/**
 * Allows Scala-like stripMargin
 *
 * Parameters are applied implicitly via ES2015.
 *
 * @example
 * // returns "The Number is:\n    100\nThanks for playing!"
 * let num = 100
 * let result = stripMargin`The Number is:
 *         |    ${num}
 *         |Thanks for playing!`
 * get the gist: https://gist.github.com/jimschubert/06fea56a6d2a1e7fdbc2
 */
export function stripMargin(template: TemplateStringsArray, ...expressions: any[]) {
    const result = template.reduce((accumulator, part, i) => {
        return accumulator + expressions[i - 1] + part;
    });

    return result.replace(/(\n|\r|\r\n)\s*\|/g, '$1');
}
