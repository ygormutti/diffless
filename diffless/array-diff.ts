import { Equal, LCS, LCSResult } from './lcs';
import {
    Change,
    ChangeLevel,
    ChangeType,
    Document,
    DocumentUri,
    Location,
    Position,
    Positioned,
    Range,
} from './model';

export interface ItemMapper<T extends Positioned> {
    (document: Document): T[];
}

class ItemWrapper<T> {
    constructor(
        readonly item: T,
        public paired: boolean = false,
    ) { }
}

const LCS_THRESHOLD = 1;

export function arrayDiff<T extends Positioned>(
    lcs: LCS,
    level: ChangeLevel,
    equal: Equal<T>,
    itemMapper: ItemMapper<T>,
    left: Document,
    right: Document,
): Change[] {
    const leftWrapped = itemMapper(left).map(wrapItem);
    const rightWrapped = itemMapper(right).map(wrapItem);
    const equalWrapped = wrapEqual(equal);
    const lcsResults: LCSResult<ItemWrapper<T>>[] = [];

    let result = lcs(equalWrapped, leftWrapped, rightWrapped);
    while (result.lcs.length > LCS_THRESHOLD) {
        pair(leftWrapped, result.leftOffset, lcs.length);
        pair(rightWrapped, result.rightOffset, lcs.length);
        lcsResults.push(result);
        result = lcs(equalWrapped, leftWrapped, rightWrapped);
    }

    let changes: Change[] = [];

    const additions = findUnpairedRanges(rightWrapped).map(
        r => new Change(level, ChangeType.Add, undefined, new Location(right.uri, r)),
    );
    changes = changes.concat(additions);

    const deletions = findUnpairedRanges(leftWrapped).map(
        r => new Change(level, ChangeType.Delete, new Location(left.uri, r), undefined),
    );
    changes = changes.concat(deletions);

    // TODO detect moves
    // const rightSortedResults = lcsResults.slice().sort((a, b) => a.rightOffset - b.rightOffset);
    // const leftSortedResults = lcsResults.slice().sort((a, b) => a.leftOffset - b.leftOffset);
    // for (let l = 0, r = 0; l < lcsResults.length;) {
    //     const leftResult = leftSortedResults[l];
    //     const rightResult = rightSortedResults[r];

    //     if (leftResult === rightResult) {
    //         ++l;
    //         ++r;
    //     } else {
    //         const leftStart = new Position()
    //         const leftLocation = new Location(left.documentUri, new Range())
    //         ++l;
    //         changes.push(new Change(level, ChangeType.Move,))
    //     }
    // }

    return changes;
}

function wrapItem<T>(item: T): ItemWrapper<T> {
    return new ItemWrapper(item);
}

function wrapEqual<T>(equal: Equal<T>): Equal<ItemWrapper<T>> {
    return (l: ItemWrapper<T>, r: ItemWrapper<T>): boolean => {
        return !r.paired && !l.paired && equal(l.item, r.item);
    };
}

function pair<T>(array: ItemWrapper<T>[], offset: number, length: number): void {
    for (let i = offset; i < (offset + length); i++) {
        const item = array[i];
        item.paired = true;
    }
}

function findUnpairedRanges<T extends Positioned>(array: ItemWrapper<T>[]): Range[] {
    const ranges = [];

    let start: Position | undefined;
    let end: Position | undefined;
    for (const wrapper of array) {
        const { paired, item: { position } } = wrapper;
        end = position;
        if (start === undefined) {
            if (paired) {
                continue;
            } else {
                start = position;
            }
        } else {
            if (paired) {
                ranges.push(new Range(start, end));
                start = undefined;
            }
        }
    }

    if (start !== undefined && end !== undefined) {
        end = new Position(end.line, end.character + 1);
        ranges.push(new Range(start, end));
    }

    return ranges;
}
