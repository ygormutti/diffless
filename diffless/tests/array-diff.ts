import { charactersDiff } from '../array-diff';
import { dynamicProgrammingLCS } from '../lcs';
import { ChangeLevel, Character, Document } from '../model';
import { stripMargin } from '../util';
import { announceHtml } from './test-util';

describe('array-diff', () => {
    describe('charactersDiff', () => {
        const diff = charactersDiff(dynamicProgrammingLCS);

        it('should detect additions and deletions correctly', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |d
                |efg`,
            );
            const right = new Document('string:right', stripMargin
                `bc
                |de
                |fgh`,
            );

            const changes = diff(left, right);
            expect(changes).toMatchSnapshot();
            announceHtml(left, right, changes, 'basic');
        });

        it('should detect moves correctly 1', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def`,
            );
            const right = new Document('string:right', stripMargin
                `def
                |abc`,
            );

            const changes = diff(left, right);
            expect(changes).toMatchSnapshot();
            announceHtml(left, right, changes, 'moves1');
        });

        it('should detect moves correctly 2', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def
                |ghi`,
            );
            const right = new Document('string:right', stripMargin
                `def
                |gabc
                |ghi`,
            );

            const changes = diff(left, right);
            expect(changes).toMatchSnapshot();
            announceHtml(left, right, changes, 'moves2');

        });

        it('should detect moves correctly 3', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def
                |ghi`,
            );
            const right = new Document('string:right', stripMargin
                `abc
                |ghi
                |def`,
            );

            const changes = diff(left, right);
            expect(changes).toMatchSnapshot();
            announceHtml(left, right, changes, 'moves3');
        });
    });
});
