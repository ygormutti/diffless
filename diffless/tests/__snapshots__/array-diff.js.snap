// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array-diff charactersDiff should detect additions and deletions correctly 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [2:2; 3:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [3:3; 3:4),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [1:1; 1:2),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:2; 3:2),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
]
`;

exports[`array-diff charactersDiff should detect moves correctly 1 1`] = `
Array [
  Object {
    "left": Location {
      "range": [1:1; 2:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [2:1; 3:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array-diff charactersDiff should detect moves correctly 2 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [3:1; 3:2),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [1:1; 2:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [2:2; 3:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
  Object {
    "left": Location {
      "range": [3:2; 4:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [3:2; 4:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array-diff charactersDiff should detect moves correctly 3 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [1:4; 2:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [3:1; 3:4),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [2:1; 2:4),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;
